src/main/java/org/opensearch/security/resources/ResourceAccessHandler.java

```diff
@@ -0,0 +1,401 @@
+    }
+
+    /**
+     * Checks if the given resource is shared with the specified entities.
+     *
+     * @param document  The ResourceSharing document to check.
+     * @param recipient The recipient entity
+     * @param entities  The set of entities to check for sharing.
+     * @param actionGroups The set of action groups to check for sharing.
+     *
+     * @return True if the resource is shared with the entities, false otherwise.
+     */
+    private boolean isSharedWithEntity(ResourceSharing document, Recipient recipient, Set<String> entities, Set<String> actionGroups) {
+        for (String entity : entities) {
+            if (checkSharing(document, recipient, entity, actionGroups)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Checks if the given resource is shared with everyone, i.e. the entity list is "*"
+     *
+     * @param document The ResourceSharing document to check.
+     * @return True if the resource is shared with everyone, false otherwise.
+     */
+    private boolean isSharedWithEveryone(ResourceSharing document) {
+        return document.getShareWith() != null
+            && document.getShareWith()
+                .getSharedWithActionGroups()
+                .stream()
+                .anyMatch(sharedWithActionGroup -> sharedWithActionGroup.getActionGroup().equals("*"));
+    }
+
+    /**
+     * Checks if the given resource is shared with the specified entity.
+     *
+     * @param document   The ResourceSharing document to check.
+     * @param recipient  The recipient entity
+     * @param entity     The entity to check for sharing.
+     * @return True if the resource is shared with the entity, false otherwise.
+     */
+    private boolean checkSharing(ResourceSharing document, Recipient recipient, String entity, Set<String> actionGroups) {
+        // Check if document is private (only visible to owner and super-admins)
+        if (document.getShareWith() == null) {
+            return false;
+        }
+
+        return document.getShareWith()
+            .getSharedWithActionGroups()
+            .stream()
+            .filter(sharedWithActionGroup -> actionGroups.contains(sharedWithActionGroup.getActionGroup()))
+            .findFirst()
+            .map(sharedWithActionGroup -> {
+                SharedWithActionGroup.ActionGroupRecipients aGs = sharedWithActionGroup.getSharedWithPerActionGroup();
+                Map<Recipient, Set<String>> recipients = aGs.getRecipients();
+
+                return switch (recipient) {
+                    case Recipient.USERS, Recipient.ROLES, Recipient.BACKEND_ROLES -> recipients.getOrDefault(recipient, Set.of())
+                        .contains(entity);
+                };
+            })
+            .orElse(false); // Return false if no matching action-group is found
+    }
+
+    private void validateArguments(Object... args) {
```

Comment:
I think we should use annotations like `@NonNull` instead of having this generic method. When annotating the args in the method signature, it would have the advantage of showing up in the IDE as the arg must be NonNull instead of finding out at runtime.

Reply:
ah good point. done.